tools/kafka-producer-performance/main.go

```diff
@@ -0,0 +1,358 @@
+	result, err := sarama.ParseKafkaVersion(version)
+	if err != nil {
+		printUsageErrorAndExit(fmt.Sprintf("unknown -version: %s", version))
+	}
+	return result
+}
+
+func main() {
+	flag.Parse()
+
+	if *brokers == "" {
+		printUsageErrorAndExit("-brokers is required")
+	}
+	if *topic == "" {
+		printUsageErrorAndExit("-topic is required")
+	}
+	if *messageLoad <= 0 {
+		printUsageErrorAndExit("-message-load must be greater than 0")
+	}
+	if *messageSize <= 0 {
+		printUsageErrorAndExit("-message-size must be greater than 0")
+	}
+
+	config := sarama.NewConfig()
+
+	config.Producer.MaxMessageBytes = *maxMessageBytes
+	config.Producer.RequiredAcks = sarama.RequiredAcks(*requiredAcks)
+	config.Producer.Timeout = *timeout
+	config.Producer.Partitioner = parsePartitioner(*partitioner, *partition)
+	config.Producer.Compression = parseCompression(*compression)
+	config.Producer.Flush.Frequency = *flushFrequency
+	config.Producer.Flush.Bytes = *flushBytes
+	config.Producer.Flush.Messages = *flushMessages
+	config.Producer.Flush.MaxMessages = *flushMaxMessages
+	config.Producer.Return.Successes = true
+	config.ClientID = *clientID
+	config.ChannelBufferSize = *channelBufferSize
+	config.Version = parseVersion(*version)
+
+	if err := config.Validate(); err != nil {
+		printErrorAndExit(69, "Invalid configuration: %s", err)
+	}
+
+	// Construct -messageLoad messages of appoximately -messageSize random bytes.
+	messages := make([]*sarama.ProducerMessage, *messageLoad)
+	for i := 0; i < *messageLoad; i++ {
+		payload := make([]byte, *messageSize)
+		if _, err := rand.Read(payload); err != nil {
+			printErrorAndExit(69, "Failed to generate message payload: %s", err)
+		}
+		messages[i] = &sarama.ProducerMessage{
+			Topic:     *topic,
+			Partition: int32(*partition),
+			Value:     sarama.ByteEncoder(payload),
+		}
+	}
+
+	// Print out metrics periodically.
+	metricsDone := make(chan struct{})
+	ctx, stopMetrics := context.WithCancel(context.Background())
+	go func(ctx context.Context) {
+		t := time.Tick(5 * time.Second)
+		for {
+			select {
+			case <-t:
+				printMetrics(os.Stdout, config.MetricRegistry)
+			case <-ctx.Done():
+				metricsDone <- struct{}{}
+				return
+			}
+		}
+	}(ctx)
+
+	brokers := strings.Split(*brokers, ",")
+	if *sync {
+		runSyncProducer(config, brokers, messages, *throughput)
+	} else {
+		runAsyncProducer(config, brokers, messages, *throughput)
+	}
+
+	stopMetrics()
+	<-metricsDone
+	close(metricsDone)
+
+	// Print final metrics.
+	printMetrics(os.Stdout, config.MetricRegistry)
+}
+
+func runAsyncProducer(config *sarama.Config, brokers []string,
+	messages []*sarama.ProducerMessage, throughput int) {
+	producer, err := sarama.NewAsyncProducer(brokers, config)
+	if err != nil {
+		printErrorAndExit(69, "Failed to create producer: %s", err)
+	}
+	defer func() {
+		if err := producer.Close(); err != nil {
+			printErrorAndExit(69, "Failed to close producer: %s", err)
+		}
+	}()
+
+	messagesDone := make(chan struct{})
```

Comment:
you can pass context to these function and then exit on `context.Done()` from go routine.

Reply:
Do you mean passing the context from the main routine to `runAsyncProducer` or instantiating a new context in this function and replacing `messagesDone` with it?  The first case is not desirable because we don't want to exit from this function until this goroutine has finished (i.e. all the messages have been successfully sent, or an error occurred). The second case is really just a matter of style; we can wait on `context.Done` at the end of the function or this `messagesDone` channel.  I don't have a huge preference here.